clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2185,9 +2179,143 @@ void handleFailure(Throwable throwable) {
+                        future.completeExceptionally(error.exception());
+                        pendingTopics.remove(topicName);
+                        if (responseCursor != null && responseCursor.topicName().equals(topicName)) {
+                            responseCursor = null;
+                        }
+                        continue;
+                    }
+
+                    TopicDescription currentTopicDescription = getTopicDescriptionFromDescribeTopicsResponseTopic(topic, nodes);
+
+                    if (partiallyFinishedTopicDescription != null && partiallyFinishedTopicDescription.name().equals(topicName)) {
+                        // Add the partitions for the cursor topic of the previous batch.
+                        partiallyFinishedTopicDescription.partitions().addAll(currentTopicDescription.partitions());
+                        continue;
+                    }
+
+                    if (responseCursor != null && responseCursor.topicName().equals(topicName)) {
+                        // In the same batch of result, it may need to handle the partitions for the previous cursor
+                        // topic and the current cursor topic. Cache the result in the nextTopicDescription.
+                        nextTopicDescription = currentTopicDescription;
+                        continue;
+                    }
+
+                    pendingTopics.remove(topicName);
+                    future.complete(currentTopicDescription);
+                }
+
+                if (partiallyFinishedTopicDescription != null &&
+                        (responseCursor == null || !responseCursor.topicName().equals(partiallyFinishedTopicDescription.name()))) {
+                    // We can't simply check nextTopicDescription != null here to close the partiallyFinishedTopicDescription.
+                    // Because the responseCursor topic may not show in the response.
+                    String topicName = partiallyFinishedTopicDescription.name();
+                    topicFutures.get(topicName).complete(partiallyFinishedTopicDescription);
+                    pendingTopics.remove(topicName);
+                    partiallyFinishedTopicDescription = null;
+                }
+                if (nextTopicDescription != null) {
+                    partiallyFinishedTopicDescription = nextTopicDescription;
+                }
+
+                if (!pendingTopics.isEmpty()) {
+                    runnable.call(this, time.milliseconds());
+                }
+            }
+
+            @Override
+            boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                final long now = time.milliseconds();
+                log.warn("The DescribeTopicPartitions API is not supported, using Metadata API to describe topics.");
+                runnable.call(generateDescribeTopicsCallWithMetadataAPI(topicNamesList, topicFutures, options, now), now);
+                return false;
+            }
```

Comment:
If I remember correctly, this will get called before we actually send off the request, right? (Assuming the broker doesn't support the new API)

Reply:
Correct, the client will directly return the UnsupportedVersion error to the _Call_ if the DescribeTopicPartitions API is not in the ApiVersionResponse